File: Secure Code Review/XSS/Java/Domxss-fix.java
import javax.servlet.ServletException;
import javax.servlet.annotation.WebServlet;
import javax.servlet.http.*;
import java.io.IOException;

@WebServlet("/search")
public class SecureDOMXSSServlet extends HttpServlet {
    protected void doGet(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
        response.setContentType("text/html");

        response.getWriter().println("<html>");
        response.getWriter().println("<head><title>Search</title></head>");
        response.getWriter().println("<body>");
        response.getWriter().println("<h1>Search</h1>");
        response.getWriter().println("<input type='text' id='query' />");
        response.getWriter().println("<button onclick='doSearch()'>Search</button>");
        response.getWriter().println("<div id='result'></div>");

        // Secure JavaScript with proper encoding
        response.getWriter().println("<script>");
        response.getWriter().println("function doSearch() {");
        response.getWriter().println("  var params = new URLSearchParams(window.location.search);");
        response.getWriter().println("  var query = params.get('query');");
        response.getWriter().println("  if (query) {");
        response.getWriter().println("    var sanitizedQuery = document.createTextNode(query);"); // Use text nodes instead of innerHTML
        response.getWriter().println("    var resultDiv = document.getElementById('result');");
        response.getWriter().println("    resultDiv.textContent = 'Results for: ';");
        response.getWriter().println("    resultDiv.appendChild(sanitizedQuery);");
        response.getWriter().println("  }");
        response.getWriter().println("}");
        response.getWriter().println("</script>");

        response.getWriter().println("</body>");
        response.getWriter().println("</html>");
    }
}






























/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
.............................................................................................................................................................................................

Explanation of Fixes

    Use textContent or document.createTextNode():
        Replacing innerHTML with textContent or appending sanitized text nodes ensures the input is treated as plain text rather than HTML, neutralizing potential XSS payloads.

    Sanitize Input:
        While textContent and createTextNode inherently escape input, validating and sanitizing the input before use adds an additional layer of protection.

    Avoid Parsing Query Strings Manually:
        Use the URLSearchParams API, which safely extracts query parameters without risking manual parsing errors.

    Example of Safe Output:
        Input: <script>alert('XSS')</script>
        Rendered in Browser: Results for: <script>alert('XSS')</script> (as plain text, not executed).

Best Practices to Prevent DOM-Based XSS

    Minimize Dynamic Content Injection:
        Avoid injecting untrusted user input into the DOM. Use safe methods like textContent.

    Use Encoding Libraries:
        Consider libraries like OWASP Java Encoder or DOMPurify for additional sanitization if user input must be used.

    Implement Content Security Policy (CSP):
        Configure a CSP header to restrict script execution:

    response.setHeader("Content-Security-Policy", "default-src 'self'; script-src 'self';");

Validate Input:

    Validate user input on both the server side and client side to ensure it meets expected formats.

File: Secure Code Review/XSS/.Net/domxss-fix.cs
#Frontend

<!DOCTYPE html>
<html>
<head>
    <title>User Profile</title>
</head>
<body>
    <h1>User Profile</h1>
    <p id="message"></p>

    <script>
        // Retrieve the `name` parameter from the URL
        const urlParams = new URLSearchParams(window.location.search);
        const name = urlParams.get("name");

        // Sanitize user input by escaping special characters
        function sanitize(input) {
            const div = document.createElement("div");
            div.appendChild(document.createTextNode(input));
            return div.innerHTML;
        }

        // Safely insert sanitized input into the DOM
        document.getElementById("message").textContent = "Welcome, " + (name ? sanitize(name) : "Guest") + "!";
    </script>
</body>
</html>




















.........................................................................................................................................
/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

Explanation of Fixes

    Input Sanitization:
        The sanitize function creates a new DOM element, adds the user input as a text node, and retrieves the escaped output.
        This ensures special characters like <, >, and " are safely encoded, preventing malicious code execution.

    Use of textContent:
        Instead of innerHTML, use textContent to inject user-provided data into the DOM. textContent inserts text as plain content, so it cannot execute HTML or JavaScript.

    Safe Output Example:
        Malicious Input: <script>alert('XSS')</script>
        Rendered Output: Welcome, &lt;script&gt;alert('XSS')&lt;/script&gt;
        The browser displays the input as plain text without executing it.

Additional Security Measures

    Avoid Unnecessary Client-Side Rendering:
        If possible, process user input on the server side and send sanitized data to the client.

    Content Security Policy (CSP):
        Implement a CSP header to block inline scripts and restrict script execution:

    Response.Headers.Add("Content-Security-Policy", "default-src 'self'; script-src 'self';");

Validation:

    Validate user input on the server side before it is ever sent to the client.

Security Libraries:

    Use security libraries like DOMPurify for robust sanitization if the application requires rendering HTML from untrusted sources.      




